### PR TITLE
🍒[cxx-interop] Lower macOS deployment target version for the Cxx module

### PR DESCRIPTION
**Pull request by egorzhdan**
**Explanation**: The Cxx module does not rely on any modern OS features. It can be deployed to older macOS versions. Having a high deployment target blocks certain usage workflows of the Cxx module.
**Scope**: All of the code changes are on a C++ interop only code path.
**Risk**: Low, only takes effect when C++ interop is enabled.
**Testing**: Covered by LIT tests.

Original PR: https://github.com/apple/swift/pull/69512

rdar://117699474

### CODE REVIEW

File: stdlib/public/Cxx/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,8 @@ add_swift_target_library(swiftCxx STATIC NO_LINK_NAME IS_STDLIB IS_SWIFT_ONLY IS
     # For functionality that depends on the C++ stdlib, use C++ stdlib overlay (`swiftstd` module).
     -Xcc -nostdinc++
 
+    DEPLOYMENT_VERSION_OSX ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_OSX}
```
Comment (DougGregor) [inline stdlib/public/Cxx/CMakeLists.txt:26]: It seems like this needs to be done for all Apple OS's, as we do for the various compatibility libraries, e.g.,
```
  DEPLOYMENT_VERSION_OSX ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_OSX}
  DEPLOYMENT_VERSION_IOS ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_IOS}
  DEPLOYMENT_VERSION_TVOS ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_TVOS}
  DEPLOYMENT_VERSION_WATCHOS ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_WATCHOS}

  MACCATALYST_BUILD_FLAVOR "zippered"
```

File: stdlib/public/Cxx/std/CMakeLists.txt [hunk 2]
```diff
@@ -49,6 +49,8 @@ add_swift_target_library(swiftCxxStdlib STATIC NO_LINK_NAME IS_STDLIB IS_SWIFT_O
     SWIFT_COMPILE_FLAGS_LINUX
     ${SWIFT_SDK_LINUX_CXX_OVERLAY_SWIFT_COMPILE_FLAGS}
 
+    DEPLOYMENT_VERSION_OSX ${COMPATIBILITY_MINIMUM_DEPLOYMENT_VERSION_OSX}
```
Comment (DougGregor) [inline stdlib/public/Cxx/std/CMakeLists.txt:52]: Same comment here about all Apple platforms

File: stdlib/cmake/modules/SwiftSource.cmake [hunk 3]
```diff
@@ -452,6 +457,7 @@ function(_compile_swift_files
       "${SWIFT_STDLIB_BUILD_TYPE}"
       "${SWIFT_STDLIB_ASSERTIONS}"
       swift_flags
+      DEPLOYMENT_VERSION_OSX ${SWIFTFILE_DEPLOYMENT_VERSION_OSX}
```
Comment (DougGregor) [inline stdlib/cmake/modules/SwiftSource.cmake:460]: I'm surprised that any of these changes are needed, given that `add_swift_target_library` is already being used with these `DEPLOYMENT_VERSION_*` options elsewhere. Are you sure these changes here are necessary?

**Review by DougGregor — CHANGES_REQUESTED** [review]

File: stdlib/cmake/modules/SwiftSource.cmake (hunk 3 above)
Comment (egorzhdan) [reply stdlib/cmake/modules/SwiftSource.cmake:460]: `add_swift_target_library` currently only applies these options for C/C++ source files. Here we're making it apply `DEPLOYMENT_VERSION_*` for Swift modules as well.

**Comment by egorzhdan** [comment]
@swift-ci please test

File: stdlib/public/Cxx/CMakeLists.txt (hunk 1 above)
Comment (egorzhdan) [reply stdlib/public/Cxx/CMakeLists.txt:26]: Done, I added other Apple platforms here

**Review by DougGregor — APPROVED** [review]
Looks good now, thank you!

**Pull request merged**